Changes from patch comments, debugged.

ndev2 opened the pull request.

Commit:
Changes from patch comments, debugged. Checkpatch.pl checked.

cjdoucette commented on lib/utils.c:
```diff
@@ -39,6 +39,54 @@
 
+int read_prop(char *dev, char *prop, long *value)
+{
+	char fname[41], buf[80], *endp;
```
Is there a reason you replaced `IFNAMSIZ+25` with `41`? I think we should keep it the way it was.

cjdoucette replied:
Actually, I think even `IFNAMSIZ+25` is kind of broken -- really long property names (like `gro_flush_timeout`) will cause the filename to be something like `/sys/class/net/enp3s0f2/gro_flush_timeout` to be over this limit. Let's just make it `char fname[128]`.

cjdoucette commented on lib/utils.c:
```diff
@@ -39,6 +39,54 @@
+	FILE *fp;
+	long result;
+
+	sprintf(fname, "/sys/class/net/%s/%s", dev, prop);
```
Let's make this `sprintf()` safer by changing it to `snprintf()`, with the second parameter being `sizeof(fname)`.

cjdoucette replied:
Let's also catch any errors that happen from `snprintf()` like we did in `parse_percent_rate()`.

cjdoucette commented on lib/utils.c:
```diff
@@ -39,6 +39,54 @@
+	sprintf(fname, "/sys/class/net/%s/%s", dev, prop);
+	fp = fopen(fname, "r");
+	if (fp == NULL) {
+		if (strcasecmp(prop, "tun_flags"))
```
The reviewer said to remove this special case, so let's delete this line.

cjdoucette commented on lib/utils.c:
```diff
@@ -39,6 +39,54 @@
+	if (fp == NULL) {
+		if (strcasecmp(prop, "tun_flags"))
+			fprintf(stderr, "open %s: %s\n", fname,
+				strerror(errno));
```
Once we delete the `if` statement that wraps this `fprintf()`, let's move it back on level of indentation. Also, make the error message read "fopen" instead of "open."

cjdoucette commented on lib/utils.c:
```diff
@@ -39,6 +39,54 @@
+	len = fread(buf, sizeof(buf), 1, fp);
+	fclose(fp);
+	if (len < 0) {
+		fprintf(stderr, "read %s: %s", fname, strerror(errno));
```
Make this "fread" instead of "read."

cjdoucette commented on lib/utils.c:
```diff
@@ -39,6 +39,54 @@
+		return -1;
+	}
+
+	len = fread(buf, sizeof(buf), 1, fp);
```
The return value of `fread()` is different than `read()` -- try reading the manual page and see if you can come up with the correct way to treat the return value. Note that the manual specifies that we need to use `feof()` and `ferror()` to get it right:

       On  success,  fread()  and  fwrite() return the number of items read or
       written.  This number equals the number of bytes transferred only  when
       size  is 1.  If an error occurs, or the end of the file is reached, the
       return value is a short item count (or zero).

       fread() does not distinguish between end-of-file and error, and callers
       must use feof(3) and ferror(3) to determine which occurred.

It may help to look for uses of `fread()` in this repository or in the kernel or any other codebase to see if you can get it right.

cjdoucette commented on lib/utils.c:
```diff
@@ -39,6 +39,54 @@
+		return -1;
+	}
+
+	result = strtoul(buf, &endp, 0);
```
I'll need to look more into the manual to get the return value here right, but in your next rebase if you have time you should try it as well. Looking at examples of uses of `strtoul()` in this repository or the kernel may help.

ndev2 commented:
Hi Cody, I've modified the code incorporating the changes, I've addressed all the possible error returns of  `strtoul()` as well (hopefully, 95% sure) . Hope the code is good, thanks.

cjdoucette commented on lib/utils.c:
```diff
@@ -39,6 +39,56 @@
+
+	ret = snprintf(fname, sizeof(fname), "/sys/class/net/%s/%s", dev, prop);
+	if (ret <= 0 || ret >= sizeof(fname))
+		goto out;
```
I think we could use a different error message here (instead of the one at the `out` label), because the failure in this case is not that we're not able to parse the file. It's that we're not able to form the filename.

ndev2 replied:
I realised that `snprintf()` could very rarely go wrong here, and in that case: similar code in the repository where its used, `snprintf()` doesn't check for errors, and in the rare cases that it does, it just returns -1 (because we really can't be sure what went wrong in this case). Do let me know how we proceed, thanks.

cjdoucette commented on lib/utils.c:
```diff
@@ -39,6 +39,56 @@
+		fprintf(stderr, "Unexpected EOF during fread %s: %s", fname, strerror(errno));
+	}
+
+	fclose(fp);
```
**Note: this comment is about how to handle `fread()` correctly, but we won't be using `fread()` because the reviewer suggested using `fgets()`. So read this for understanding, but not to change the code.**

I know I said to look at other examples of `fread()` in this repository for how to deal with the return value, but it looks like they're not handling it correctly in another file that I found, so let's just stick with the manual.

The prototype of this function is:

    size_t fread(void *ptr, size_t size, size_t nmemb, FILE *stream);

Notice that the return type is `size_t`, which is an unsigned value. That means that the return value can never be negative, which makes sense based on what the manual says:

> On success, fread() and fwrite() return the number of items read or written... If an error occurs, or the end of the file is reached, the return value is a short item count (or zero).

In other words, the return value is always going to be greater than or equal to zero, but an error occurs whenever the return value is not equal to the third parameter.

Here's a closer, but still not correct, example of handling the return value: http://en.cppreference.com/w/c/io/fread#Example

So we should have something like:

    size_t count;
    ...
    count = fread(buf, sizeof(buf), 1, fp);
    if (count != 1) {
    	if (feof(fp))
    		fprintf(stderr, "Error reading %s: unexpected end of file\n", fname);
    	else if (ferror(fp))
    		fprintf(stderr, "Error reading %s\n", fname);
    	fclose(fp);
    	return -1;
    }
    fclose(fp);

Notice that the manual page for `fread()` doesn't say anything about setting `errno`, so we shouldn't use `strerror(errno)` in the case of an error here. The page above tries to use `perror()` as if `errno` had been set, but it's wrong.

cjdoucette replied:
All of the above is a good lesson in figuring out return values, but looking at the feedback from the reviewer, it looks like he wants us to use `fgets()` instead of `fread()`. That makes sense because we're trying to read an entire line from the file.

The return value of `fgets()` is a lot easier, so I'll let you take a shot at it.

ndev2 replied:
Thanks Cody! 👍  :)

cjdoucette commented on lib/utils.c:
```diff
@@ -39,6 +39,56 @@
+	}
+
+	fclose(fp);
+	result = strtoul(buf, &endp, 0);
```
I don't think the handling below is exactly right. I think we just want:

    result = strtoul(buf, &endp, 0);
    if (buf == endp || *endp) {
    	fprintf("value \"%s\" in file %s is not a number\n", buf, fname);
    	goto out;
    }
    if (result == ULONG_MAX && errno == ERANGE) {
    	fprintf(stderr, "strtoul %s: %s", fname, strerror(errno));
    	goto out;
    }

ndev2 replied:
But shouldn't we do something when  nothing is interpreted by `strtoul()` , which happens when the file has an empty string (cases like wlan0 or bridge devices, etc)? The `result == 0` case takes care of that. Let me know what you think.

ndev2 replied:
But `fgets()` would catch that case now, so  I don't think we need `result == 0` case anymore, will make the change

ndev2 commented:
Hi Cody, I've made the changes, but now, we don't explicitly say what error occurred for `snprintf()`, we just return -1, do let me know if you'd like me to change that, but let me know the error message to print to, thanks. Also, hope the other changes are good, do let me know, thanks.

cjdoucette commented on lib/utils.c:
```diff
@@ -39,6 +39,46 @@
+
+	ret = snprintf(fname, sizeof(fname), "/sys/class/net/%s/%s", dev, prop);
+	if (ret <= 0 || ret >= sizeof(fname))
+                return -1;
```
Let's add an error message here: `fprintf(stderr, "could not build pathname for property\n");`

cjdoucette commented on lib/utils.c:
```diff
@@ -39,6 +39,46 @@
+	}
+
+	if (!fgets(buf, sizeof(buf), fp))
+		goto out;
```
Before the `goto out`, let's close the file handle: `fclose(fp);`.

cjdoucette reviewed (commented):
Just a couple of comments. Let me know once you rebase, I'll do a quick review, and then we should be good to go.

ndev2 commented:
Hi Cody, I've made the changes, I've also run `checpatch.pl` and it only gives the `over 80 characters limit` warning for line number 70, other than that everything is good. Do let me know if we're good to push the patch, thanks.

cjdoucette commented:
OK, if that's the case then on line 70 just move the `buf` and `fname` parameters onto their own line (indenting by an extra tab) so that checkpatch doesn't complain. Then feel free to update your patch to the mailing list so that they can review it again!

ndev2 committed:
Merge branch 'master' into patch_2

ndev2 commented:
Hi Cody,
                  Our check in loc 70 seems to be creating a problem. I'm
not sure of the check here, because of the 2nd condition in the OR
statement. I'm thinking if we replace this check with : if (result == 0 &&
errno == EINVAL )... it would work (that's what the 1st condition in the OR
statement does, but I guess this would be safer because it evaluates only
the return value). Or we could just remove this check (because fgets()
would catch it anyways). I've attached a screenshot of the test, do let me
know what you think. Thanks.

Thanks for your time,
Nishanth

cjdoucette commented:
It looks like the merging commit introduces some changes that is causing some compilation problems:

https://github.com/ndev2/iproute2-final/pull/4/commits/57b1654f1a748bd21508d7d5dd8a267d6ef855a7

Could you squash all these patches down to just be under one commit, "B.W. limits can now be specified in %"?

ndev2 commented:
Hi Cody,
               Yeah, there were some merge errors initially(when I merged
the pull request) because my keyboard spazzed out accidentally, but I
resolved them and made a final commit (the latest master repo on GitHub
doesn't have any merge conflicts) , I tested the final code on the VM
('server' in the screenshot) and that's the error that came up. The command
runs but  we shouldn't be getting an error here , so I traced the bug back
to loc 70. Do let me know what you think, thanks.


Thanks for your time,
Nishanth

On Nov 9, 2017 13:53, "Cody Doucette" <notifications@github.com> wrote:

> It looks like the merging commit introduces some changes that is causing
> some compilation problems:
>
> 57b1654
> <https://github.com/ndev2/iproute2-final/commit/57b1654f1a748bd21508d7d5dd8a267d6ef855a7>
>
> Could you squash all these patches down to just be under one commit, "B.W.
> limits can now be specified in %"?
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/ndev2/iproute2-final/pull/4#issuecomment-343303645>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AbWxghsTgSmjdIe2KLoCyPrGVpSn5OZEks5s03RGgaJpZM4QV8NV>
> .
>

ndev2 commented:
Oh sorry Cody, I was responding from a phone and thought this was in the other email thread.
